dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/PostCacheEventPublisher.java

```diff
@@ -114,6 +157,46 @@ private void publishMessage( Class<?> realClass, String message )
         }
     }
 
+    private Serializable getDataValueId( Object entity )
+    {
+        DataValue dataValue = (DataValue) entity;
+
+        long dataElementId = dataValue.getDataElement().getId();
+        long periodId = dataValue.getPeriod().getId();
+        long organisationUnitId = dataValue.getSource().getId();
+        long categoryOptionComboId = dataValue.getAttributeOptionCombo().getId();
+        long attributeOptionComboId = dataValue.getAttributeOptionCombo().getId();
+
+        return new DataValue.DataValueId(dataElementId, periodId, organisationUnitId, categoryOptionComboId, attributeOptionComboId);
+    }
+
+    private Serializable getTrackedEntityAttributeValueId( Object entity )
+    {
+        TrackedEntityAttributeValue trackedEntityAttributeValue = (TrackedEntityAttributeValue) entity;
+
+        long trackedEntityAttributeId = trackedEntityAttributeValue.getAttribute().getId();
+        long entityInstanceId = trackedEntityAttributeValue.getEntityInstance().getId();
+
+        TrackedEntityAttribute trackedEntityAttribute = trackedEntityAttributeService.getTrackedEntityAttribute(
+            trackedEntityAttributeId );
+        TrackedEntityInstance entityInstance = trackedEntityInstanceService.getTrackedEntityInstance(
+            entityInstanceId );
+
+        return trackedEntityAttribute.getUid() + ";" + entityInstance.getUid();
+    }
+
+    private Serializable getCompleteDataSetRegistrationId( Object entity )
+    {
+        CompleteDataSetRegistration completeDataSetRegistration = (CompleteDataSetRegistration) entity;
+
+        long dataSetId = completeDataSetRegistration.getDataSet().getId();
+        long periodId = completeDataSetRegistration.getPeriod().getId();
+        long organisationUnitId = completeDataSetRegistration.getSource().getId();
+        long categoryOptionComboId = completeDataSetRegistration.getAttributeOptionCombo().getId();
+
+        return dataSetId + ";" + periodId + ";" + organisationUnitId + ";" + categoryOptionComboId;
```

Comment:
I think this part can be written better by joining e.g. with `String.join(";", List.of(dataSetId, periodId, ..))`.

Reply:
This does not work without having to cast to charseq